src/sentry/tasks/relocation.py

```diff
@@ -95,5 +137,333 @@ def uploading_complete(uuid: str) -> None:
+    need to validate that none of these are mutated during import.
+
+    This function is meant to be idempotent, and should be retried with an exponential backoff.
+    """
+
+    relocation: Optional[Relocation]
+    attempts_left: int
+    (relocation, attempts_left) = start_task(
+        uuid=uuid,
+        step=Relocation.Step.PREPROCESSING,
+        task=OrderedTask.PREPROCESSING_COLLIDING_USERS,
+        allowed_task_attempts=MAX_FAST_TASK_ATTEMPTS,
+    )
+    if relocation is None:
+        return
+
+    with retry_task_or_fail_relocation(
+        relocation,
+        OrderedTask.PREPROCESSING_COLLIDING_USERS,
+        attempts_left,
+        ERR_PREPROCESSING_INTERNAL,
+    ):
+        fp = BytesIO()
+        export_in_user_scope(
+            fp,
+            encrypt_with=BytesIO(get_public_key_using_gcp_kms(DEFAULT_CRYPTO_KEY_VERSION)),
+            user_filter=set(relocation.want_usernames),
+        )
+        fp.seek(0)
+        kind = RelocationFile.Kind.COLLIDING_USERS_VALIDATION_DATA
+        file = File.objects.create(name=kind.to_filename("tar"), type=RELOCATION_FILE_TYPE)
+        file.putfile(fp, blob_size=RELOCATION_BLOB_SIZE, logger=logger)
+        RelocationFile.objects.create(
+            relocation=relocation,
+            file=file,
+            kind=kind.value,
+        )
+
+        preprocessing_complete.delay(uuid)
+
+
+@instrumented_task(
+    name="sentry.relocation.preprocessing_complete",
+    queue="relocation",
+    max_retries=MAX_FAST_TASK_RETRIES,
+    retry_backoff=RETRY_BACKOFF,
+    retry_backoff_jitter=True,
+    soft_time_limit=PREPROCESSING_TIME_LIMIT,
+    silo_mode=SiloMode.REGION,
+)
+def preprocessing_complete(uuid: str) -> None:
+    """
+    Creates a "composite object" from the uploaded tarball, which could have many pieces. Because
+    creating a composite object in this manner is a synchronous operation, we don't need a follow-up
+    step confirming success.
+
+    This function is meant to be idempotent, and should be retried with an exponential backoff.
+    """
+
+    relocation: Optional[Relocation]
+    attempts_left: int
+    (relocation, attempts_left) = start_task(
+        uuid=uuid,
+        step=Relocation.Step.PREPROCESSING,
+        task=OrderedTask.PREPROCESSING_COMPLETE,
+        allowed_task_attempts=MAX_FAST_TASK_ATTEMPTS,
+    )
+    if relocation is None:
+        return
+
+    with retry_task_or_fail_relocation(
+        relocation,
+        OrderedTask.PREPROCESSING_COMPLETE,
+        attempts_left,
+        ERR_PREPROCESSING_INTERNAL,
+    ):
+        storage = get_storage()
+        for kind in RELOCATION_FILES_TO_BE_VALIDATED:
+            raw_relocation_file = (
+                RelocationFile.objects.filter(
+                    relocation=relocation,
+                    kind=kind.value,
+                )
+                .select_related("file")
+                .prefetch_related("file__blobs")
```

Comment:
Not familiar with this, but just wanted to confirm that the double underscore is intentional here for `file__blobs`

Reply:
It's Django's magic notation for `fetch through another model` - in this case, I'm saying "get me the model referred to by the foreign key field `blob` on the `File` model".